numba/core/types/functions.py

```diff
@@ -375,33 +375,44 @@ def get_call_type(self, context, args, kws):
         template = self.template(context)
         literal_e = None
         nonliteral_e = None
+        out = None
 
-
-        # Try with Literal
-        try:
-            out = template.apply(args, kws)
-        except Exception as exc:
-            if isinstance(exc, errors.ForceLiteralArg):
-                raise exc
-            literal_e = exc
-            out = None
-
-        # if the unliteral_args and unliteral_kws are the same as the literal
-        # ones, set up to not bother retrying
-        unliteral_args = tuple([unliteral(a) for a in args])
-        unliteral_kws = {k: unliteral(v) for k, v in kws.items()}
-        skip = unliteral_args == args and kws == unliteral_kws
-
-        # If the above template application failed and the non-literal args are
-        # different to the literal ones, try again with literals rewritten as
-        # non-literals
-        if not skip and out is None:
-            try:
-                out = template.apply(unliteral_args, unliteral_kws)
-            except Exception as exc:
-                if isinstance(exc, errors.ForceLiteralArg):
-                    raise exc
-                nonliteral_e = exc
+        choice = [True, False] if template.prefer_literal else [False, True]
+        for uselit in choice:
+            if uselit:
+                # Try with Literal
+                try:
+                    out = template.apply(args, kws)
+                except Exception as exc:
+                    if isinstance(exc, errors.ForceLiteralArg):
+                        raise exc
+                    literal_e = exc
+                    out = None
+                else:
+                    break
+            else:
+                # if the unliteral_args and unliteral_kws are the same as the literal
+                # ones, set up to not bother retrying
+                unliteral_args = tuple([unliteral(a) for a in args])
+                unliteral_kws = {k: unliteral(v) for k, v in kws.items()}
```

Comment:
`BaseFunction::get_call_type` uses `_unlit_non_poison` https://github.com/numba/numba/pull/6058/files#diff-758157deb494d50a6d6c93447b370a05R227-R234 want to reuse here?

Reply:
good pt

Reply:
done in https://github.com/numba/numba/pull/6058/commits/efaa1a7455b2af6d9edd0dd98bd9cb3f4707c826